lib/helpers.ts

```diff
@@ -189,12 +189,19 @@ function getSampleAndPatientData(
         .filter((f) => !!f) as Entity[];
 
     const diagnosis = biospecimen
-        .map(
-            (s) =>
-                diagnosisByHTANParticipantID[s.HTANParentID] as
-                    | Entity
-                    | undefined
-        )
+        .map((s) => {
+            // HTANParentID can be both participant or biospecimen, so keep
+            // going up the tree until participant is found.
+            let HTANParentID = s.HTANParentID;
+            while (HTANParentID in biospecimenByHTANBiospecimenID) {
```

Comment:
would this be more clear if straight up testing for participant identity instead of testing to see it is NOT a biospecimen?

Reply:
I had the same thought. one counterpoint - this is a bit safer because it's testing membership in biospecimenByHTAN.. before accessing that same map

Reply:
if the data is incorrect it could happen that `HTANParentID` is never in `diagnosisByHTANParticipantID`

will add a check for that

Reply:
how about that (see below). Added some logging. Not sure if that's your recommended logging strategy. But it works, also logs known issues with HTAPP

Reply:
I think logging is a good idea. I just noticed something else - the map is indexed by HTANBiospecimenID, but we're using HTANParentID to index into it. a bit confusing. is there another name that we can use that covers both?

Reply:
yeah I was considering to `s/HTANParentID/biospecimenOrParticipantID`, but ofc further down the if/else logic you know which one it is, so the naming makes less sense. In some ways `HTANParentID` is actually kind of clear in that it could be both (tho it's maybe not obvious that it'll e.g. never by a FileID). The naming of the maps themselves make sense to me because they are indexed by only one type of ID

Reply:
the only thing thats not clear is that an HTANParentID can also be an HTANParticipantID or an HTANBiospecimenID. like, its not clear that those are from the same set of ids. im not sure what the solution is, I agree renaming the map to be so verbose may actually be more confusing. I think the issue to resolve is how to clarify that these ids come from the same set *sometimes*